src/blockstore/hashmap_db.zig

```diff
@@ -0,0 +1,254 @@
+        defer self.allocator.free(key_bytes);
+        var map = self.maps[cf_index];
+
+        self.transaction_lock.lockShared();
+        defer self.transaction_lock.unlockShared();
+        map.lock.lockShared();
+        defer map.lock.unlockShared();
+
+        const val_bytes = map.getPreLocked(key_bytes) orelse return null;
+
+        const ret = try self.allocator.alloc(u8, val_bytes.len);
+        @memcpy(ret, val_bytes);
+        return .{
+            .allocator = self.allocator,
+            .data = ret,
+        };
+    }
+
+    pub fn delete(
+        self: *Self,
+        comptime cf: ColumnFamily,
+        comptime cf_index: usize,
+        key: cf.Key,
+    ) !void {
+        const key_bytes = try cf.key().serializeAlloc(self.allocator, key);
+        defer self.allocator.free(key_bytes);
+        self.transaction_lock.lockShared();
+        defer self.transaction_lock.unlockShared();
+        _ = self.maps[cf_index].delete(self.allocator, key_bytes);
+    }
+
+    pub fn writeBatch(self: *Self) MapBatch {
+        return .{
+            .allocator = self.allocator,
+            .instructions = .{},
+        };
+    }
+
+    pub fn initBatch(self: *Self) MapBatch {
+        return MapBatch.init(self.allocator);
+    }
+
+    /// Atomicity may be violated if there is insufficient
+    /// memory to complete a PUT.
+    pub fn commit(self: *Self, batch: MapBatch) !void {
+        self.transaction_lock.lock();
+        defer self.transaction_lock.unlock();
+
+        for (batch.instructions.items) |ix| {
+            switch (ix) {
+                .put => |put_ix| {
+                    const cf_index, const key, const value = put_ix;
+                    try self.maps[cf_index].put(key, value);
+                },
+                .delete => |delete_ix| {
+                    const cf_index, const key = delete_ix;
+                    self.maps[cf_index].delete(batch.allocator, key);
+                },
+            }
+        }
+    }
+};
+
+pub const MapBatch = struct {
+    allocator: Allocator,
+    instructions: std.ArrayListUnmanaged(Instruction),
+
+    const Instruction = union(enum) {
+        put: struct { usize, []const u8, []const u8 },
+        delete: struct { usize, []const u8 },
+    };
+
+    const Self = @This();
+
+    pub fn init(allocator: Allocator) Self {
+        return .{
+            .allocator = allocator,
+            .instructions = std.ArrayListUnmanaged(Instruction){},
+        };
+    }
+
+    pub fn deinit(self: Self) void {
+        self.instructions.deinit(self.allocator);
+    }
+
+    pub fn put(
+        self: *Self,
+        comptime cf: ColumnFamily,
+        comptime cf_index: usize,
+        key: cf.Key,
+        value: cf.Value,
+    ) !void {
+        return try self.instructions.append(self.allocator, .{ .put = .{
+            cf_index,
+            try cf.key().serializeAlloc(self.allocator, key),
+            try cf.value().serializeAlloc(self.allocator, value),
```

Comment:
Careful with these kinds of nested expressions, this is hiding a leak in the error path. If the first try succeeds, but the second fails, the first is leaked. If both of these trys succeed, but the try on the append fails, that's two leaks. Best to split these kind of things up into assignments with respective `errdefer free()`s.

Reply:
0aa78ac7f94501f2203178748849b47fe16ec7be

I'm afraid we may be missing errdefers throughout the codebase. It's easy to overlook these negative cases.

Reply:
Yeah, it's likely we are - there are ways to snuff out edge case memory leaks we could/should look to utilizing, namely `std.testing.checkAllAllocationFailures`. We could create both general and targeted fuzz tests for that.
But it's still a win when we can catch them early in review like this.